nimare/meta/cbmr.py

```diff
@@ -0,0 +1,293 @@
+                            all_group_study_id[group] = group_study_id.unique().tolist()
+                elif isinstance(self.group_names, list):
+                    not_exist_group_names = [group for group in self.group_names if group not in dataset_annotations.columns]
+                    if len(not_exist_group_names) > 0:
+                        raise ValueError("group_names: {} does not exist in the dataset".format(not_exist_group_names))
+                    uniq_group_splits = dataset_annotations[self.group_names].drop_duplicates().values.tolist()
+                    for group in uniq_group_splits:
+                        group_study_id_bool = (dataset_annotations[self.group_names] == group).all(axis=1)
+                        group_study_id = dataset_annotations.loc[group_study_id_bool]['study_id']
+                        all_group_study_id['_'.join(group)] = group_study_id.unique().tolist()
+                self.inputs_['all_group_study_id'] = all_group_study_id
+                # collect studywise moderators if specficed
+                if hasattr(self, "moderators"):
+                    all_group_moderators = dict()
+                    for group in all_group_study_id.keys():
+                        df_group = dataset_annotations.loc[dataset_annotations['study_id'].isin(all_group_study_id[group])] 
+                        group_moderators = np.stack([df_group[moderator_name] for moderator_name in self.moderators], axis=1)
+                        group_moderators = group_moderators.astype(np.float64)
+                        group_moderators = self._standardize_moderators(group_moderators)
+                        all_group_moderators[group] = group_moderators
+                    self.inputs_["all_group_moderators"] = all_group_moderators
+                # Calculate IJK matrix indices for target mask
+                # Mask space is assumed to be the same as the Dataset's space
+                # These indices are used directly by any KernelTransformer
+                all_foci_per_voxel, all_foci_per_study = dict(), dict()
+                for group in all_group_study_id.keys():
+                    group_study_id = all_group_study_id[group]
+                    group_coordinates = dataset.coordinates.loc[dataset.coordinates['study_id'].isin(group_study_id)] 
+                    # group-wise foci coordinates
+                    group_xyz = group_coordinates[['x', 'y', 'z']].values
+                    group_ijk = mm2vox(group_xyz, mask_img.affine)
+                    group_foci_idx = vox2idx(group_ijk, mask_img._dataobj)
+                    # number of foci per voxel/study
+                    n_group_study = len(group_study_id)
+                    masker_voxels = np.sum(mask_img._dataobj).astype(int)
+                    group_foci_per_voxel = np.zeros((masker_voxels, 1))
+                    group_foci_per_voxel[group_foci_idx, :] += 1
```

Comment:
@yifan0330 this seems to be the only place where `vox2idx` is used. From what I can tell, IJK here (`group_ijk`) is relative to the full 91x109x91 image space, rather than the reduced one you referred to in the call, which simplifies things somewhat.

One option, which might be slightly slower, would be to create a 3D array of zeros with the same shape as the mask and then loop through the IJK arrays, adding to the 3D array as you go. Then you just create a new image object from the 3D array and apply the masker to that.

```python
group_foci_per_voxel = np.zeros(mask_img.shape, dtype=int)
for ijk in group_ijk:
    group_foci_per_voxel[ijk[0], ijk[1], ijk[2]] += 1

# will not work with maskers that aren't NiftiMaskers
group_foci_per_voxel = nib.Nifti1Image(group_foci_per_voxel, mask_img.affine, mask_img.header)
group_foci_per_voxel = self.masker.transform(group_foci_per_voxel)
```